source/s3_client.c

```diff
@@ -1397,6 +1397,9 @@ static void s_s3_client_acquired_retry_token(
 
     AWS_ASSERT(client->vtable->acquire_http_connection);
 
+    /* client needs to be kept alive until s_s3_client_on_acquire_http_connection completes */
+    aws_s3_client_acquire(client);
```

Comment:
Discussed offline:
Feel like this is a life time problem between aws_s3_endpoint and aws_s3_client. We need to dig more into it, but probably not a blocker. We should add a TODO here in the code.